### PR TITLE
Implement Enumerable methods

### PR DESCRIPTION
**Pull request by iliebabcenco**
There we have:

- my_each
- my_each_with_index
- my_select
- my_all?
- my_any?
- my_none?
- my_count
- my_map
- my_inject
- multiply_els

### CODE REVIEW

File: enumerable.rb [hunk 1]
```diff
@@ -0,0 +1,115 @@
+module Enumerable
+  def my_each
+    to_a.length.times { |i| yield self[i] }
+    self
+  end
```
Comment (ElijahTobs) [inline enumerable.rb]: ## my_each

*  [x] returns an Enumerator if no block is given
```ruby
array.my_each
```
Expected: Enumerator
got: LocalJumpError: no block given (yield)

* [x] when a block is given when self is a range returns the range itself after calling the given block once for each element in self
```ruby
range.my_each(&block)
```
got: NoMethodError:
       undefined method `[]' for 5..50:Range
- [x] when a block is given when self is a range it calls the given block once for each element in self
```ruby  
my_each_output = ''
block = proc { |num| my_each_output += num.to_s }
range.my_each(&block)
```
got NameError:
undefined local variable or method `length' for 5..50:Range

- [x] when a block is given when self is a hash it calls the given block once for each element in self
```ruby
my_each_output = ''
block = proc { |num| my_each_output += num.to_s }
hash.my_each(&block)
```
expected: "[:a, 1][:b, 2][:c, 3][:d, 4][:e, 5]"
got: ""

File: enumerable.rb [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+
+  def my_each_with_index
+    to_a.my_each { |each| yield each, index(each) }
+    self
+  end
```
Comment (ElijahTobs) [inline enumerable.rb]: ## my_each_with_index

 - [x] returns an enumerator if no block is given
```ruby
array.my_each_with_index
```
Expected: Enumerator
got:LocalJumpError:
       no block given (yield)

- [x] when a block is given when self is an array behaves like #each_with_index calls the given block once for each element in self
```ruby
my_each_with_index_output
```
Expected: "Num: 3, idx: 0\nNum: 5, idx: 1\nNum: 5, idx: 2\nNum: 7, idx: 3\nNum: 1, idx: 4\nNum: 1, idx: 5\nNum:...: 8, idx: 94\nNum: 8, idx: 95\nNum: 4, idx: 96\nNum: 4, idx: 97\nNum: 2, idx: 98\nNum: 5, idx: 99\n"
got: "Num: 3, idx: 0\nNum: 5, idx: 1\nNum: 5, idx: 1\nNum: 7, idx: 3\nNum: 1, idx: 4\nNum: 1, idx: 4\nNum:...Num: 8, idx: 7\nNum: 8, idx: 7\nNum: 4, idx: 23\nNum: 4, idx: 23\nNum: 2, idx: 12\nNum: 5, idx: 1\n"

- [x] when a block is given when self is a range returns the range itself after calling the given block once for each element in self
```ruby
range.my_each_with_index(&block)
```
got: NoMethodError:
       undefined method `index' for 5..50:Range

- [x] when a block is given when self is a range behaves like #each_with_index calls the given block once for each element in self
```ruby
enum.my_each_with_index(&block)
```
got: NoMethodError:
       undefined method `index' for 5..50:Range

- [x] when a block is given when self is a hash behaves like #each_with_index calls the given block once for each element in self
```ruby
my_each_with_index_output
```
expected: "Num: [:a, 1], idx: 0\nNum: [:b, 2], idx: 1\nNum: [:c, 3], idx: 2\nNum: [:d, 4], idx: 3\nNum: [:e, 5], idx: 4\n"
            got: "Num: [:a, 1], idx: \nNum: [:b, 2], idx: \nNum: [:c, 3], idx: \nNum: [:d, 4], idx: \nNum: [:e, 5], idx: \n"

File: enumerable.rb [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+
+  def my_select
+    array = []
+    to_a.my_each { |each| array.push(each) if yield each }
+    array
+  end
```
Comment (ElijahTobs) [inline enumerable.rb]: ## my_select

- [x] returns an enumerator if no block is given
```ruby
array.my_select
```
Expected: Enumerator
got: LocalJumpError:
        no block given (yield)

File: enumerable.rb [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+        return false unless yield each
+      elsif parameter.instance_of? Class
+        return false unless each.is_a? parameter
+      else
+        return false unless each
+        return false unless parameter === each
+      end
+    end
+    true
+  end
```
Comment (ElijahTobs) [inline enumerable.rb]: ## my_all
- [x] when no block or argument is given returns true when none of the collection members are false or nil
```ruby
true_array.my_all?
```
expected true
             got false

File: enumerable.rb [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+        return true if yield each
+      elsif parameter.instance_of? Class
+        return true if each.is_a? parameter
+      else
+        return true if each == true
+        return true if parameter === each
+      end
+    end
+    false
+  end
```
Comment (ElijahTobs) [inline enumerable.rb]: ## my_any
- [x] when no block or argument is given returns false if at least one of the collection is not true
```ruby
false_array.my_any?
```
expected false
             got true

File: enumerable.rb [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+    return to_enum unless block_given? || !proc.nil?
+
+    array = []
+    if proc.nil?
+      to_a.my_each { |each| array.push(yield each) }
+    else
+      to_a.my_each { |_each| array.push(proc.call(to_a[i])) }
+    end
+    array
+  end
```
Comment (ElijahTobs) [inline enumerable.rb]: ## my_map
- [x] executes only the proc when both a block and a proc are given and returns a new array with the results of running the proc once for every element in enum.
```ruby
array.my_map(my_proc) { |num| num < 10 }
```
got: NameError:
        undefined local variable or method `i' for #<Array:0x0000557be3bc2500>

File: enumerable.rb [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+        return first_param
+      end
+    elsif symbol?(first_param)
+      to_a.my_each { |each| result = result.nil? ? each : result.send(first_param, each) }
+    else
+      to_a.my_each { |each| first_param = first_param.send(second_param, each) }
+      return first_param
+    end
+    result
+  end
```
Comment (ElijahTobs) [inline enumerable.rb]: ## my_inject
- [x] raises a "LocalJumpError" when no block or argument is given
```ruby
array.my_inject
```
expected: LocalJumpError
 got: #<TypeError: nil is not a symbol nor a string> with backtrace:

**Review by ElijahTobs — CHANGES_REQUESTED** [review]
# Status: Changes Required :recycle: 

## Hi @iliebabcenco and @Henryhaulka :wave: 

You've done well so far :+1:  and I commend you for that :clap: 
However, there are still some changes to be made before this milestone can be approved.

### Variables used during testing

```ruby
ARRAY_SIZE = 100
LOWEST_VALUE = 0
HIGHEST_VALUE = 9

array =  Array.new(ARRAY_SIZE) { rand(LOWEST_VALUE...HIGHEST_VALUE) }
block = proc { |num| num < (LOWEST_VALUE + HIGHEST_VALUE) / 2 } 
words = %w[dog door rod blade]
range = Range.new(5, 50)
hash = { a: 1, b: 2, c: 3, d: 4, e: 5 } 
numbers = [1, 2i, 3.14]
array_clone = array.clone
```


## REQUIRED :recycle: 

* [x] Check below :point_down: 


## Happy Coding :computer: 

_Feel free to reach out to me on slack_ **`@Elijah Ayandokun`**
  **Elijah O. Ayandokun**
[Github](https://github.com/ElijahTobs) | [LinkedIn](https://www.linkedin.com/in/ayandokunelijah/) | [Twitter](https://twitter.com/ElijahTobs)

**Review by jubaan — CHANGES_REQUESTED** [review]
[![Microverse][microverse-shield]][microverse-url]
<div>
<h1 align="center"> ⚠️ 🛠️ ⚠️ 🛠️ Changes Required 🛠️ ⚠️ 🛠️ ⚠️ </h1>
<br />
<p><b>"Fail fast, fail often"</b> and don't let this discourage you, because <b>only by failing do we learn to succeed</b>.</p>
<details open="open">
  <summary>Review Summary</summary>
  <ol>
    <li><a href="#congrats"> 💯 Awesome work 👏🏼 👏🏼 </a></li>
    <li><a href="#mandatory"> ⚠️ Things to Improve </a> 🚨 🚨 </li>
    <li><a href="#optional"> 💡Nice to have and other comments ✍🏼 ✍🏼 </a></li>
    <li><a href="#contact"> 🧔🏼 About your Reviewer </a> (you don't know it... but we are just automated bots 🤖) </li>
    <li><a href="#help"> 🆘 <b> 😭 Feeling stuck? or Need help</b>? 🙏🏼 🙏🏼 </a></li>
  </ol>
</details>
</div>

<h3 id="congrats"> 1. :100: Awesome work</h3>
<details close="close">
<summary> [click to open] Big or little achievements, but achievements at the end. <b>Feel proud that you're learning and improving everyday.</b></summary>



</details>

<hr>
<h3 id="mandatory"> 2. ⚠️ Things to Improve (or the To-Do list to get the project approved) </h3> 
<details close="close">
<summary> [click to open] Working on these observations will help you advance and get closer to the goal 🎯, postpone them or do them half-ways and they'll haunt you 👻 👻... FOREVER!! (It will feel like that, trust me). </summary>


- [ ] enumerables #my_all? when no block or argument is given returns false when one of the collection members are false or nil.



</details>
<p><b>NOTE</b>This list is or could be divided in what is show in this comment and inline comments posted below.<p>
<hr>

<h3 id="optional"> 3. 💡 Nice to have and other comments </h3>
<details close="close">
<summary> [click to open] </h3></summary>


- [x] Nothing more to add.


</details>

<h3 id="contact"> 4. 🧔🏼 About your Reviewer </h3>
<details close="close">
<summary> [click to open] Say "hello world!" on Slack: @jubaan</summary>
<br />
<div align="center">
  <a src="https://www.github.com/jubaan">
    Julio Añoveros
  </a>
  <p>
    Julio Añoveros | @jubaan</br></br>
    <b>Full-stack Web Developer</b> and</br>
    🤖 Code Reviewer at <a href="https://www.microverse.org/?grsf=kfm2ej">Microverse</a>:rocket:
  </p>
</div>
</details>

<h3 id="help"> 5. 🆘 <b> 😭 Feeling stuck? or Need help</b>? 🙏🏼 🙏🏼</h3>
<details close="close">
  <summary> [click to open] <b>Feel stuck?</b> </summary>
  <br>
  <p> First breath... just do it. Take 15-20 min away from the computer, stretch, meditate, do some exercise, drink water, or do whatever is necessary to distract your mind from the obstacle.</p>
  <p><b>Sometimes you just need to get away to see things with a new perspective.</b></p>
</details>

<details close="close">
  <summary> [click to open] <b>Need help?</b> </summary>
  <br>
  <p> Send me a DM in <a href="https://www.microverse.org/?grsf=kfm2ej">Microverse</a> Students Slack, you'll find me as @jubaan. If I'm not available there are other Code Reviewers you can contact. They'll help as well.</p>
  <p>Remember that <a href="https://www.microverse.org/?grsf=kfm2ej">Microverse</a> has other channels where you can get an answer to your questions:</p>
  <ol>
    <li>The <b>students channel 🧑‍🎓 </b> in slack,</li>
    <li>your <b>teammates 👥 </b>,</li>
    <li>your <b>mentor 👨🏼 👩🏼 </b>, and</li>
    <li><b>Student Success 🏫 </b>.</li>
  </ol>
  <p><b> DON'T hesitate to ask! </b></p>
</details>
<p>If all of the above doesn't help... get a <a href="https://rubberduckdebugging.com/">rubber duck</a> ... it would be great to own one 😄. </p>

<h2 align="center"> Keep up the great work :muscle: and Happy Coding :smiley: :computer:!! </h1>

[microverse-shield]: https://img.shields.io/badge/-microverse-%236F23FF?style=for-the-badge
[microverse-url]: https://www.microverse.org/?grsf=kfm2ej
[bootstrap-shield]: https://img.shields.io/badge/Bootstrap-white?style=for-the-badge&logo=bootstrap
[bootstrap-url]: https://www.microverse.org/?grsf=kfm2ej

**Review by andresporras3423 — CHANGES_REQUESTED** [review]
### Status: Changes Required :recycle:
**Hello Henry Onu & Ilie Babcenco** 
Amazing job :muscle:  :muscle:
Your methods pass all the tests :clap: :clap: :clap:  
Just one last issue to correct before you can go ahead:

- [x] Don't edit rubocop config file
 
Please request another review once you are ready! :+1: 
 
Happy coding :relaxed:
 
Oscar Andrés Russi Porras

**Review by GzimAsani — APPROVED** [review]
# STATUS: Project Approved :green_circle: 

## Hello team, you have done great job so far, congrats and well done with this project :+1: :muscle: 
I am sorry that it was not mentioned earlier but you still have 4 errors coming from my_none? and my_all?. Even though they are marked as optional I highly suggest that you complete them:
- [OPTIONAL] #my_all? when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern
- [OPTIONAL] when a pattern other than Regex or a Class is given returns false if any of the collection does not match the pattern
- [OPTIONAL] #my_none? when a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern
- [OPTIONAL] #my_none? when a pattern other than Regex or a Class is given returns false only if one of the collection matches the pattern

### Happy coding :computer: :desktop_computer:

**Pull request merged**